rustup-init.sh

```diff
@@ -400,30 +402,50 @@ downloader() {
         get_ciphersuites_for_curl
         _ciphersuites="$RETVAL"
         if [ -n "$_ciphersuites" ]; then
-            curl --proto '=https' --tlsv1.2 --ciphers "$_ciphersuites" --silent --show-error --fail --location "$1" --output "$2"
+            _err=$(curl --proto '=https' --tlsv1.2 --ciphers "$_ciphersuites" --silent --show-error --fail --location "$1" --output "$2" 2>&1)
+            _status=$?
         else
             echo "Warning: Not enforcing strong cipher suites for TLS, this is potentially less secure"
             if ! check_help_for "$3" curl --proto --tlsv1.2; then
                 echo "Warning: Not enforcing TLS v1.2, this is potentially less secure"
-                curl --silent --show-error --fail --location "$1" --output "$2"
+                _err=$(curl --silent --show-error --fail --location "$1" --output "$2" 2>&1)
+                _status=$?
             else
-                curl --proto '=https' --tlsv1.2 --silent --show-error --fail --location "$1" --output "$2"
+                _err=$(curl --proto '=https' --tlsv1.2 --silent --show-error --fail --location "$1" --output "$2" 2>&1)
+                _status=$?
             fi
         fi
+        if [ -n "$_err" ]; then
+            echo "$_err" >&2
+            if echo "$_err" | grep -q 404$; then
+                err "installer for platform '$3' not found, this may be unsupported"
+            fi
+        fi
+        return $_status
     elif [ "$_dld" = wget ]; then
         get_ciphersuites_for_wget
         _ciphersuites="$RETVAL"
         if [ -n "$_ciphersuites" ]; then
-            wget --https-only --secure-protocol=TLSv1_2 --ciphers "$_ciphersuites" "$1" -O "$2"
+            _err=$(wget --https-only --secure-protocol=TLSv1_2 --ciphers "$_ciphersuites" "$1" -O "$2" 2>&1)
+            _status=$?
         else
             echo "Warning: Not enforcing strong cipher suites for TLS, this is potentially less secure"
             if ! check_help_for "$3" wget --https-only --secure-protocol; then
                 echo "Warning: Not enforcing TLS v1.2, this is potentially less secure"
-                wget "$1" -O "$2"
+                _err=$(wget "$1" -O "$2" 2>&1)
+                _status=$?
             else
-                wget --https-only --secure-protocol=TLSv1_2 "$1" -O "$2"
+                _err=$(wget --https-only --secure-protocol=TLSv1_2 "$1" -O "$2" 2>&1)
+                _status=$?
+            fi
+        fi
+        if [ -n "$_err" ]; then
+            echo "$_err" >&2
+            if echo "$_err" | grep -q ' 404 Not Found$'; then
```

Comment:
I would grep for `ERROR 404`.

Reply:
Would that also suffer from the localisation problem I pointed out?

Reply:
Yes, see https://github.com/rust-lang/rustup/pull/2435#discussion_r460714244. `错误 404` for simplified chinese and `エラー 404` for japanese so I bet other localization are there too.

Reply:
Ugh. sorry for the poor suggestion. Given wgets lack of a programmable interface I'm actually inclined to suggest dropping it as a supported downloader. It would make what we support simpler and there are no platforms I know of where curl is not available but wget is.

Reply:
No, I do encounter some platforms where wget is available but curl isn't, I don't think we need to drop support for something we already support.

Reply:
I don't mind if we give slightly less good errors on platforms where we have to fall back to wget.

Reply:
Out of interest, what platforms are these? We added this support only 2 years ago in a PR with no attached bug report, that said 'linux systems have this often', but rustup had been around for many years at that point and there are literally no bug reports of it not working at that time. I'm not worried about less-good-errors, I'm worried about less-correct-behaviour.

Reply:
Or maybe easier, why not just search for `-w 404` to match exactly 404 for both curl and wget? Will that be incorrect?

Reply:
Maybe? What would that look like?

Reply:
`curl $args | grep -w 404`

Reply:
`404 Not Found` is not localized, did you try `env LANG=zh_CN.UTF-8 wget --https-only --secure-protocol=TLSv1_2 "https://static.rust-lang.org/rustup/dist/armv7-unknown-linux-musleabihf/rustup-init" -O "/tmp/rustup-init"` or something similar?

The line after that however is localized, `错误 404：Not Found。`

Reply:
IIRC we thought that the 'Not Found' was part of the AWS web frontend and thus subject to potential change, however if you think it's part of wget then I'm prepared to let it slide.

Reply:
I believe it's part of wget, I saw the same error message for different sites.

Reply:
Aha, okay